File: src/II083.cpp
#include <iostream>
#include <vector>
using namespace std;

// 基于交换的回溯法
class Solution {
  void getRes(vector<int>& nums, vector<vector<int>>& res, int i_start = 0) {
    int n = nums.size();
    if (i_start == n) {
      res.push_back(nums);
    }
    for (int i = i_start; i < n; ++i) {
      swap(nums[i_start], nums[i]);
      getRes(nums, res, i_start + 1);
      swap(nums[i_start], nums[i]);
    }
  }

public:
  vector<vector<int>> permute(vector<int>& nums) {
    vector<vector<int>> res;
    getRes(nums, res);
    return res;
  }
};

File: src/II083.rs
struct Solution;

impl Solution {
    pub fn permute(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
        fn next_permutation(nums: &mut Vec<i32>) -> bool {
            if nums.len() <= 1 {
                return false;
            }
            let mut i = nums.len() - 2;
            while nums[i] >= nums[i + 1] {
                if i == 0 {
                    return false;
                }
                i -= 1;
            }
            let mut j = nums.len() - 1;
            while nums[j] <= nums[i] {
                j -= 1;
            }
            nums.swap(i, j);
            i += 1;
            j = nums.len() - 1;
            while i < j {
                nums.swap(i, j);
                i += 1;
                j -= 1;
            }
            true
        }
        let mut res = vec![];
        nums.sort();
        res.push(nums.clone());
        while next_permutation(&mut nums) {
            res.push(nums.clone());
        }
        res
    }
}
